Fixed: Image preview should not zoom on first click if unfocused

MartinBrathen opened the pull request:
@mjbvz I believe this should solve issue #81877

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/vscode?pullRequest=82074) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/vscode?pullRequest=82074) before we can accept your contribution.<br/><br/>:x: MartinBrathen [sign now](https://cla.opensource.microsoft.com/microsoft/vscode?pullRequest=82074)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/vscode?pullRequest=82074) it.</sub>

MartinBrathen commented:
Pressed the "Update branch" button, now test are not passing, will try to fix somehow.

mjbvz commented on extensions/image-preview/media/main.js:
```diff
@@ -70,6 +70,8 @@
 	let altPressed = false;
 	let hasLoadedImage = false;
+	let isActive = true;
+	let canClickToZoom = false;
```
Do we need both of these or can they be combined?

mjbvz commented on extensions/image-preview/media/main.js:
```diff
@@ -152,36 +173,53 @@
 	});
 
-	container.addEventListener('click', (/** @type {MouseEvent} */ e) => {
+	container.addEventListener('mousedown', (/** @type {MouseEvent} */ e) => {
```
Also do we need both mousedown and `click`? I think we should be able to handle everything in `click`?

MartinBrathen replied:
```canClickToZoom``` and the ```'mousedown'``` event handler where added because the ```'click'``` event is fired when mouse button is lifted, causing a zoom on first click if the user does not release the mouse button before ```isActive``` is set to true. These changes ensures that only clicks that are initiated after focus is acquired results in zooming.

I think this change is good because the outcome of the first click will seem inconsistent otherwise. I will try to get the same effect without the use of ```canClickToZoom``` and the ```'mousedown'``` event handler, but I'm not sure I will succeed.

mjbvz commented on extensions/image-preview/src/preview.ts:
```diff
@@ -120,6 +120,7 @@ export class Preview extends Disposable {
 			}
 			this._previewState = PreviewState.Visible;
 		}
+		this.webviewEditor.webview.postMessage({ type: 'setActive', value: this.webviewEditor.active });
```
Instead of requiring message passing, also see if we can use `document.hasFocus()`  inside the webview. I think this may track what we want.

MartinBrathen replied:
Hi,

I have experimented a bit with ```document.hasFocus()``` and the ```'focus'``` event. ```document.hasFocus()``` is always set to true before any ```MouseEvent``` is fired, this makes it so that we can not use it to check if click was done without focus, same goes for the ```'focus'``` event. This can be solved by setting a state variable in ```focus``` event handler that blockes the first click after focus. However, this introduces a bug where if focus is gained without the use of a click, the user will be forced to click twice in order to zoom. I have not found any way to gain information about what caused the ```FocusEvent```, which would be needed to solve this bug.

When using message passing instead, the ```'message'``` event handler will be triggered after the ```'mousedown'``` event handler, but before the ```'click'``` event handler, if focus was gained through a click. This is useful, because it allows us to identify what caused the gain of focus, and to handle each case accordingly. In my solution the ```'mousedown'``` event handler changes the state, to allow a zoom to be executed only in the case where a click is initiated after focus was gained. For this this reason I will mark all as resolved.

MartinBrathen commented:
I removed an unnecessary state variable.

In this [branch](https://github.com/MartinBrathen/vscode/blob/test/extensions/image-preview/media/main.js) I removed the ```'mousedown'``` listener and the usage of message passing as well. Sadly it introduces a bug where the user has to click twice if the image preview is not focused by clicking on the preview. I do not currently know of a way to check how the preview was focused.

mjbvz commented:
Thanks! Will be in the next insiders build and is scheduled to go out with VS Code 1.40

MartinBrathen commented:
Cool! Thank you :)
